packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx

```diff
@@ -0,0 +1,220 @@
+  currentCoordinates.x !== nextCoordinates.x || currentCoordinates.y !== nextCoordinates.y;
+
+type OwnerState = { classes: DataGridProProcessedProps['classes'] };
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    rowDragging: ['row--dragging'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+export const rowReorderStateInitializer: GridStateInitializer = (state) => ({
+  ...state,
+  rowReorder: { dragRow: '' },
+});
+
+/**
+ * Only available in DataGridPro
+ * @requires useGridRows (method)
+ */
+export const useGridRowReorder = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<DataGridProProcessedProps, 'disableRowReorder' | 'onRowOrderChange' | 'classes'>,
+): void => {
+  const logger = useGridLogger(apiRef, 'useGridRowReorder');
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const treeDepth = useGridSelector(apiRef, gridRowTreeDepthSelector);
+  const dragRowNode = React.useRef<HTMLElement | null>(null);
+  const originRowIndex = React.useRef<number | null>(null);
+  const removeDnDStylesTimeout = React.useRef<any>();
+  const cursorPosition = React.useRef<CursorCoordinates>({
+    x: 0,
+    y: 0,
+  });
+  const ownerState = { classes: props.classes };
+  const classes = useUtilityClasses(ownerState);
+
+  React.useEffect(() => {
+    return () => {
+      clearTimeout(removeDnDStylesTimeout.current);
+    };
+  }, []);
+
+  // TODO: remove sortModel check once row reorder is sorting compatible
+  // remove treeDepth once row reorder is tree compatible
+  const isRowReorderDisabled = React.useCallback((): boolean => {
+    return props.disableRowReorder || !!sortModel.length || treeDepth !== 1;
+  }, [props.disableRowReorder, sortModel, treeDepth]);
+
+  const handleDragStart = React.useCallback<GridEventListener<GridEvents.rowDragStart>>(
+    (params, event) => {
+      if (isRowReorderDisabled()) {
+        return;
+      }
+
+      logger.debug(`Start dragging row ${params.id}`);
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      dragRowNode.current = event.currentTarget;
+      dragRowNode.current.classList.add(classes.rowDragging);
+
+      apiRef.current.setState((state) => ({
+        ...state,
+        rowReorder: { ...state.rowReorder, dragRow: params.id },
+      }));
+      apiRef.current.forceUpdate();
+
+      removeDnDStylesTimeout.current = setTimeout(() => {
+        dragRowNode.current!.classList.remove(classes.rowDragging);
+      });
+
+      originRowIndex.current = apiRef.current.getRowIndex(params.id);
+    },
+    [isRowReorderDisabled, classes.rowDragging, logger, apiRef],
+  );
+
+  const handleDragEnter = React.useCallback<
+    GridEventListener<GridEvents.cellDragEnter | GridEvents.rowDragEnter>
+  >((params, event) => {
+    event.preventDefault();
```

Comment:
Why do we need this listener? For the column reorder I remember it was added to prevent a conflict with React DnD. But the listener available in the column reorder prevents the event even when there's no column being dragged so this listener here could be removed, I think.